### PR TITLE
Add EXIF Metadata Extraction Support (Part One)

### PR DESCRIPTION
**Pull request by Incharajayaram**
## 📝 Description

This PR adds EXIF metadata extraction support to kornia-io, specifically focused on reading the EXIF Orientation tag (1-8) from image files without requiring full image decoding. This is the first part of a two-part feature implementation.

**Fixes/Relates to:** #639 

## 🛠️ Changes Made
- [x] Add `ImageMetadata` struct to hold EXIF orientation data (`exif_orientation: Option<u16>`)
- [x] Implement `read_image_metadata(path)` function for efficient metadata extraction
- [x] Add minimal EXIF parser that extracts Orientation tag (0x0112) without decoding pixels
- [x] Support JPEG, PNG, and TIFF file formats
- [x] Add `little_exif = "0.6"` dependency to `Cargo.toml`
- [x] Export `metadata` module from `lib.rs`
- [x] Add 3 comprehensive unit tests for metadata extraction

### Files Modified
- `crates/kornia-io/Cargo.toml` - Added `little_exif` dependency
- `crates/kornia-io/src/lib.rs` - Exported metadata module

### Files Created
- `crates/kornia-io/src/metadata.rs` - Core implementation (~350 lines)

---

### Unit Tests (3 new tests)
- [x] **test_missing_exif_returns_none**: Verifies that images without EXIF orientation tags return `None`
- [x] **test_write_and_read_orientation_values**: Tests reading specific orientation values (1, 6, 8) from embedded EXIF data
- [x] **test_all_orientation_values**: Comprehensive test covering all 8 valid EXIF orientation values (1-8)

### Test Results
All 25 tests passing (3 new + 22 existing):

```
running 25 tests
test metadata::tests::test_all_orientation_values ... ok
test metadata::tests::test_missing_exif_returns_none ... ok
test metadata::tests::test_write_and_read_orientation_values ... ok
[... 22 existing tests ...]

test result: ok. 25 passed; 0 failed; 0 ignored; 0 measured
```

### Manual Verification
- Tested with real JPEG files from repo (`tests/data/dog.jpeg`)
- Embedded EXIF orientation tags using `little_exif` crate
- Verified correct parsing of all 8 orientation values
- Verified error handling for invalid orientations (9-255, 0)

### Edge Cases Handled
- Images without EXIF data → returns `None`
- Invalid orientation values (0, 9-255) → returns `None`
- Corrupted EXIF blocks → gracefully returns `None`
- Missing files → returns `IoError::FileError`

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [ ] 🟢 **No AI used.**
- [x] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [x] I am assigned to the linked issue (required before PR submission)
- [x] The linked issue has been approved by a maintainer
- [x] This PR strictly implements what the linked issue describes (no scope creep)
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] (Optional) I have attached screenshots/recordings for UI changes.

---

## 💭 Additional Context
Add any other context or screenshots about the pull request here.

**Implementation Based On:**
- [little_exif crate (v0.6)](https://docs.rs/little_exif/latest/little_exif/) - Pure Rust EXIF parsing
- EXIF Specification (v2.32) - TIFF/Orientation tag structure
- Existing kornia-rs error handling patterns (`Result<T, IoError>`)

### Why Split into Two PRs?

This PR focuses on **metadata extraction only** to keep the scope minimal.  A follow-up PR 2 will add:
- Orientation transform functions
- Auto-orient readers
- Integration tests

### What's NOT Included (Reserved for PR 2)

-  `apply_exif_orientation()` function
- Flip/rotate/transpose operations
- `_with_orientation` reader variants
- Transform integration tests

### CODE REVIEW

**Comment by Incharajayaram** [comment]
@sidd-27 If this is fine, then I will proceed with the second PR once its merged

**Comment by sidd-27** [comment]
not all pixels need the reflect operation, perhaps we could separate into two paths for border and for center?
also the kernel weights and offsets could be precomputed and flattened for faster computation

File: crates/kornia-io/Cargo.toml [hunk 1]
```diff
@@ -35,6 +35,8 @@ jpeg-encoder = "0.6"
 kornia-image = { workspace = true }
 kornia-tensor = { workspace = true }
 log = { workspace = true }
+# EXIF parsing (metadata only)
```
Comment (sidd-27) [inline crates/kornia-io/Cargo.toml]: Please remove agentic comments, purely ai AI-generated code is not allowed

Comment (Incharajayaram) [reply crates/kornia-io/Cargo.toml]: Yes, I will make the changes.

**Comment by Incharajayaram** [comment]
@sidd-27 I think this comment was meant for the pyrdown PR #629? This EXIF PR only reads metadata from file headers.


> not all pixels need the reflect operation, perhaps we could separate into two paths for border and for center? also the kernel weights and offsets could be precomputed and flattened for faster computation

**Comment by sidd-27** [comment]
oh yea mb

**Comment by Incharajayaram** [comment]
@edgarriba please review this PR for the Exif orientation issue #640

File: crates/kornia-io/src/metadata.rs [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+        });
+    }
+
+    let ext = path
+        .as_ref()
+        .extension()
+        .and_then(|e| e.to_str())
+        .map(|e| e.to_lowercase());
+
+    let file_type = match ext.as_deref() {
```
Comment (edgarriba) [inline crates/kornia-io/src/metadata.rs:68]: can EXIF also extract the file type ?

File: crates/kornia-io/src/metadata.rs [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+        Some("jpg") | Some("jpeg") => Some(FileExtension::JPEG),
+        Some("tif") | Some("tiff") => Some(FileExtension::TIFF),
+        _ => None,
+    };
+
+    let metadata = file_type
+        .map(|ft| little_exif::metadata::Metadata::new_from_vec(&bytes, ft))
+        .unwrap_or_else(|| {
+            little_exif::metadata::Metadata::new_from_vec(&bytes, FileExtension::JPEG)
```
Comment (edgarriba) [inline crates/kornia-io/src/metadata.rs]: i'd just raise an unsupported type error to avoid silent bugs

File: crates/kornia-io/src/metadata.rs [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::fs;
+    use std::time::{SystemTime, UNIX_EPOCH};
+
+    fn write_temp_file(name_suffix: &str, data: &[u8]) -> std::path::PathBuf {
+        let mut path = std::env::temp_dir();
```
Comment (edgarriba) [inline crates/kornia-io/src/metadata.rs]: use `tempfile` https://github.com/kornia/kornia-rs/blob/main/examples/copper/src/main.rs#L10

File: crates/kornia-io/src/metadata.rs [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+            .as_millis();
+        path.push(format!("kornia_io_test_{}_{}.jpg", name_suffix, ts));
+        fs::write(&path, data).expect("write temp file");
+        path
+    }
+
+    #[test]
+    fn test_missing_exif_returns_none() {
+        let repo_img = std::path::Path::new("../../tests/data/dog.jpeg");
+        if !repo_img.exists() {
```
Comment (edgarriba) [inline crates/kornia-io/src/metadata.rs]: this image is always here

File: crates/kornia-io/src/metadata.rs (hunk 2 above)
Comment (Incharajayaram) [reply crates/kornia-io/src/metadata.rs:68]: no, exif doesn't store the file type. we need to know the format jpeg/tiff/png before parsing exif because each format stores exif data differently - jpeg uses app1 markers, tiff embeds it directly, png uses chunks.

**Comment by Incharajayaram** [comment]
addressed all review feedback  and CI checks passing. ready for re-review @edgarriba @sidd-27

File: crates/kornia-io/src/metadata.rs [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+
+use crate::error::IoError;
+use little_exif::filetype::FileExtension;
+
+/// Simple image metadata extracted during decoding.
+/// For now this only includes the EXIF Orientation tag (if present).
+/// See `read_image_metadata` for the helper function that extracts this.
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+pub struct ImageMetadata {
+    /// EXIF `Orientation` tag (1..8) if present, otherwise `None`.
```
Comment (sidd-27) [inline crates/kornia-io/src/metadata.rs:14]: look into the use of NonZeroU8 here instead of the option, this would compress this down to a single byte, or a regular u8

**Comment by sidd-27** [comment]
are you still interested in working on this issue?

**Comment by Incharajayaram** [comment]
yes yes, I am still interested 
Can you please review the changes I have made?

**Comment by Incharajayaram** [comment]
yes yes, I am still interested 
Can you please review the changes I have made?

**Comment by sidd-27** [comment]
> The PR description appears to follow a highly structured, generic template that is characteristic of LLM-generated text. Per the repository AI policy, please either replace it with a concise summary written in your own words

simplify the comments, no unnecessary stuff

**Comment by Incharajayaram** [comment]
@sidd-27 i think it was referring to PR description which is automated by copilot itself 
other comments which are present in the files are all simple and nothing unnecessary i think
